core/src/blockchain/blockchain.rs

```diff
@@ -212,6 +212,48 @@ impl BlockChain {
         }
     }
 
+    /// Force update the best block as the given block hash.
+    ///
+    /// This will not change the highest block chain. This means it is possible
+    pub fn update_best_block_hash(&self, batch: &mut DBTransaction, block_hash: H256) -> ImportRoute {
+        // FIXME: If it is possible, double check with the consensus engine.
+
+        assert!(self.pending_best_block_hash.read().is_none());
+        let block_detail = self.block_details(&block_hash).expect("The given hash should exist");
+        let parent_hash = block_detail.parent;
+
+        let prev_best_hash = self.best_block_hash();
+        let best_block_changed = if block_hash != prev_best_hash {
+            let route = tree_route(self, prev_best_hash, parent_hash)
+                .expect("blocks being imported always within recent history; qed");
+
+            let best_block =
+                self.block(&block_hash).expect("Best block is already imported as a branch").rlp().as_raw().to_vec();
+
+            match route.retracted.len() {
+                0 => BestBlockChanged::CanonChainAppended {
```

Comment:
`CanonChainAppended` assumes that `route.enacted.len() == 0`. Let's add the asssertion in the code.

Reply:
Changed the function name to `update_best_block_in_tendermint`, and added some assertion statements.